sympy/matrices/expressions/matpow.py

```diff
@@ -47,7 +48,7 @@ def _entry(self, i, j, **kwargs):
             # We still have a MatPow, make an explicit MatMul out of it.
             if A.exp.is_Integer and A.exp.is_positive:
                 A = MatMul(*[A.base for k in range(A.exp)])
-            elif A.exp.is_number:
+            elif isinstance(self.rows, (SYMPY_INTS, Integer)) and isinstance(self.cols, (SYMPY_INTS, Integer)):
```

Comment:
It's not good if self.rows/cols is not known to be sympified. Perhaps there should be a different attribute for getting symbolic shape vs explicit int shape.

Reply:
It checks both `int` and `Integer`, so non-sympified expressions should be OK.

Reply:
I mean that it is problematic if an attribute is not known to be sympified or not e.g.:
```python
In [6]: M1 = Matrix([[1]])

In [7]: M2 = MatrixSymbol('x', n, n)

In [8]: type(M1.shape[0])
Out[8]: int

In [9]: type(M2.shape[0])
Out[9]: sympy.core.symbol.Symbol
```
Maybe if the shape is sometimes a symbolic expression it should always be symbolic e.g. `Integer(1)` rather than `int(1)`.

Reply:
Well, I'm just keeping the code check the same way as they are on master (and according to `git blame` the checks for both `int` and `Integer` have been there since 2015).

Technically `.shape` is not part of `.args` of the explicit valued matrix, so it is not required to be sympified. OTOH, the `.shape` of `MatrixSymbol` is sympified.